pandas/_libs/tslibs/offsets.pyx

```diff
@@ -493,11 +499,28 @@ cdef int get_day_of_month(datetime other, day_opt) except? -1:
     30
 
     """
+    cdef:
+        int wkday, days_in_month
+
     if day_opt == 'start':
         return 1
-    elif day_opt == 'end':
-        return monthrange(other.year, other.month)[1]
+
+    wkday, days_in_month = monthrange(other.year, other.month)
+    if day_opt == 'end':
+        return days_in_month
+    elif day_opt == 'business_start':
+        # first business day of month
+        return get_firstbday(wkday, days_in_month)
+    elif day_opt == 'business_end':
+        # last business day of month
+        return get_lastbday(wkday, days_in_month)
+    elif is_integer_object(day_opt):
+        day = min(day_opt, days_in_month)
+    elif day_opt is None:
+        raise NotImplementedError
     else:
+        # Note: unlike `shift_month`, get_day_of_month does not
```

Comment:
can you add a test with a sub-class of BaseOffset (e.g. a dummy class) that checks this